src/agreement.rs

```diff
@@ -286,9 +302,12 @@ impl<NodeUid: Clone + Debug + Eq + Hash> Agreement<NodeUid> {
         };
 
         let b = self.estimated.unwrap();
-        self.sent_bval.insert(b);
-        self.messages
-            .push_back(AgreementMessage::BVal(self.epoch, b));
+        self.send_bval(b)?;
+        let queued_msgs = mem::replace(&mut self.incoming_queue, Vec::new());
+        for (sender_id, msg) in queued_msgs {
+            self.handle_message(&sender_id, msg)?;
```

Comment:
Here is one possible place for introducing spam control. I've no idea though how to avoid spam apart from using a set instead of `Vec` of queued messages and filtering out messages with unknown sender IDs. The latter is already done at the top level.

Reply:
That's true. Basically we just need to store per epoch: who sent us which `BVal`s and which `Aux`.

We still might want to limit the number of future epochs somehow. On the other hand, that wouldn't be strictly correct…

Reply:
I created a new issue for this: https://github.com/poanetwork/hbbft/issues/43